File: leetcode/javascript/hashing/countingElements/countingElements.js
/*
Учитывая целочисленный массив arr, подсчитайте, сколько элементов x существует так, что x + 1 также находится в arr.

Если в arr есть дубликаты, посчитайте их отдельно.
 */

/*
Input: arr = [1,2,3]
Output: 2
Explanation: 1 and 2 are counted cause 2 and 3 are in arr.
 */

/*
Input: arr = [1,1,3,3,5,5,7,7]
Output: 0
Explanation: No numbers are counted, cause there's no 2, 4, 6, or 8 in arr.
 */

/*
Input: arr = [1,3,2,3,5,0]
Output: 3
Explanation: 0, 1 and 2 are counted cause 1, 2 and 3 are in arr.
 */

/*
Input: arr = [1,1,2,2]
Output: 2
Explanation: Two 1s are counted cause 2 is in arr.
 */

/*
Input: arr = [1,1,2]
Output: 2
Explanation: Both 1s are counted because 2 is in the array.
 */

/**
 * @param {number[]} arr
 * @return {number}
 */
var countElements = function (arr) {
    let map = {}, total = 0;

    // Добавляем элементы в map
    for(let i of nums) map[i] ? map[i]++ : map[i] = 1;

    // Удаление повторяющихся элементов
    let newNums = [... new Set(nums)];

    // Если длина массива после удаления повторяющихся чисел меньше трёх, возвращаем 0;
    if(newNums.length < 3) return 0;

    // Сортируем массив newNums и удаляем первый и последний элементы.
    // для всех остальных элементов проверяем их количество на карте и добавляем в переменную total
    newNums.sort((a,b) => a-b).slice(1, newNums.length-1).forEach(num => total += map[num]);

    // Возвращаем итоговую переменную
    return total;
};

console.log(countElements([1,2,3]));

File: leetcode/go/hashing/countingElements(easy)/countingElements.go
package main

import (
	"fmt"
)

/* 1426. Counting Elements

Учитывая целочисленный массив arr, подсчитайте, сколько элементов x существует так, что x + 1 также находится в arr.
Если в arr есть дубликаты, посчитайте их отдельно.

Input: arr = [1,2,3]
Output: 2
Пояснение: 1 и 2 учитываются, потому что 2 и 3 находятся в arr.

Input: arr = [1,1,3,3,5,5,7,7]
Output: 0
Пояснение: Никакие числа не учитываются, поскольку в arr нет цифр 2, 4, 6 или 8.

Input: arr = [1,3,2,3,5,0]
Output: 3
Пояснение: 0, 1 и 2 учитываются, потому что 1, 2 и 3 находятся в arr.

Input: arr = [1,1,2,2]
Output: 2
Пояснение: Две единицы засчитываются, потому что 2 находится в arr.

Input: arr = [1,1,2]
Output: 2
Объяснение: Обе единицы учитываются, поскольку в массиве есть 2.
*/

func main() {
	fmt.Println(countElements([]int{1, 2, 3})) // 2
	fmt.Println(countElements([]int{1, 1, 2})) // 2
}

// countElements подсчитывает количество элементов x в массиве, таких что x + 1 также находится в массиве.
func countElements(arr []int) int {
	// Создаем множество (map) для быстрой проверки наличия элементов в массиве.
	elementSet := make(map[int]bool)
	for _, num := range arr {
		elementSet[num] = true
	}

	count := 0
	// Проходим по всем элементам массива и проверяем, есть ли (x + 1) в множестве.
	for _, num := range arr {
		if elementSet[num+1] {
			count++
		}
	}

	return count
}
